### ISSUE TITLE
Demo Company (US) gives NullPointer after com.fasterxml.jackson.databind.JsonMappingException

### ISSUE THREAD

**Issue by aschroder** [XeroAPI/Xero-Java#112]
Demo Company uses phone number of type 'Office'

But in the Enum there is no Office value:

<img width="1203" alt="screen shot 2018-12-14 at 4 28 48 pm" src="https://user-images.githubusercontent.com/387235/49981705-0d7d8880-ffbe-11e8-9e04-8365392ddea6.png">

The initial exception is:
```
Can not construct instance of com.xero.models.accounting.Phone$PhoneTypeEnum, problem: Unexpected value 'OFFICE'
```

then

```
com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of com.xero.models.accounting.Phone$PhoneTypeEnum, problem: Unexpected value 'OFFICE'
 at [Source: {
  "Id": "8e28bae6-e6f6-4d16-a1e5-9b8669e527e7",
  "Status": "OK",
  "ProviderName": "xxx",
  "DateTimeUTC": "\/Date(1544758095493)\/",
  "Organisations": [
    {
      "Name": "Demo Company (US)",
      "LegalName": "Demo Company (US)",
      "PaysTax": true,
      "Version": "US",
      "OrganisationType": "COMPANY",
      "BaseCurrency": "USD",
      "CountryCode": "US",
      "IsDemoCompany": true,
      "OrganisationStatus": "ACTIVE",
      "TaxNumber": "123456789",
      "FinancialYearEndDay": 31,
      "FinancialYearEndMonth": 12,
      "SalesTaxBasis": "ACCRUALS",
      "SalesTaxPeriod": "1MONTHLY",
      "DefaultSalesTax": "Tax Exclusive",
      "DefaultPurchasesTax": "No Tax",
      "PeriodLockDate": "\/Date(1280534400000+0000)\/",
      "CreatedDateUTC": "\/Date(1544062763580)\/",
      "OrganisationEntityType": "COMPANY",
      "Timezone": "PACIFICSTANDARDTIME",
      "ShortCode": "!Q5xw3",
      "OrganisationID": "1defe224-5257-48ab-9563-afc49b01c701",
      "Edition": "BUSINESS",
      "Class": "DEMO",
      "Addresses": [
        {
          "AddressType": "STREET",
          "AddressLine1": "79 Madison Ave, Fl 2",
          "City": "New York",
          "Region": "NY",
          "PostalCode": "10016",
          "Country": "USA",
          "AttentionTo": ""
        },
        {
          "AddressType": "POBOX",
          "AddressLine1": "23 Main Street",
          "AddressLine2": "Central City",
          "City": "Oaktown",
          "Region": "NY",
          "PostalCode": "12123",
          "Country": "USA",
          "AttentionTo": ""
        }
      ],
      "Phones": [
        {
          "PhoneType": "OFFICE",
          "PhoneNumber": "800 1234 5678"
        }
      ],
      "ExternalLinks": [],
      "PaymentTerms": {
        "Bills": {
          "Day": 14,
          "Type": "DAYSAFTERBILLDATE"
        },
        "Sales": {
          "Day": 14,
          "Type": "DAYSAFTERBILLDATE"
        }
      },
      "TaxNumberName": "e.g. State Tax ID"
    }
  ]
}; line: 55, column: 24] (through reference chain: com.xero.models.accounting.Organisations["Organisations"]->java.util.ArrayList[0]->com.xero.models.accounting.Organisation["Phones"]->java.util.ArrayList[0]->com.xero.models.accounting.Phone["PhoneType"])
```

**Comment by aschroder** [comment]
Workaround for the error above was to go and manually edit the company phone details.

Is there an old version of the demo company affecting this, even after fixing this exception with the workaround I get a new one, same sort of error:

```
com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of com.xero.models.accounting.Account$BankAccountTypeEnum, problem: Unexpected value ''
 at [Source: java.io.StringReader@3437cffd; line: 54, column: 26] (through reference chain: com.xero.models.accounting.Accounts["Accounts"]->java.util.ArrayList[2]->com.xero.models.accounting.Account["BankAccountType"])
```

Seems like the enum data model is out of sync with the demo data.

**Comment by SidneyAllen** [comment]
Hi @aschroder - 

Thanks for the feedback.  I'm updating the OpenAPI Spec today and will add

OFFICE for PhoneTypeEnum (though it's not a valid value for PUT/POST)
NONE and GSTONIMPORTS are fine for ReportTaxTypeEnum

But, I don't see a way to set an ENUM of empty string.  I'm defining these in the OAS.

https://github.com/XeroAPI/Xero-OpenAPI/blob/master/spec/v3/Xero_accounting_2.0.0_swagger.yaml

if I'm missing something, please let me know.

**Comment by aschroder** [comment]
@SidneyAllen Hmm I'm not 100% sure that it can be an enum, if an empty value is a valid value (I mean the empty string `""` not an enum value of `NONE`).

I think either you don't have an empty value - i.e you always have at least something, even if it is `NONE` or you don't use an enum for it in the API spec.

Could the API just not return that property at all if it's empty, so that it does not try to match an enum value for it at all? Or could the API return NONE and then you define a NONE in the spec yams file?

**Comment by SidneyAllen** [comment]
@aschroder - I could try adding  .... nullable: true  to the BankAccountType ?  

OAS 3 docs say ...

Note: nullable is not the same as an optional parameter or an empty-valued parameter. nullable means the parameter value can be null. Specific implementations may choose to map an absent or empty-valued parameter to null, but strictly speaking these are not the same thing.

**Comment by aschroder** [comment]
@SidneyAllen Yeh the Exception throws out of the JSON parser in that Jackson library:
```
com.fasterxml.jackson.databind.JsonMappingException
```

The SDK then chokes with a NullPointer but I didn't look into why, the root cause was the failure to parse.

I found each of the issues I worked around one by one, so there could be more in other entities/responses I never encountered. Each one caused the same issue - first in the Company, then the Accounts then in the TaxRates.

It should be reproducible with a call to get the accounts from a completely vanilla demo company.

I'm not sure if changing you spec away from enums is the right solution, these errors mostly seem related to the java library being generated form the spec, not the spec itself.

For what it's worth (and I know you have a lot more considerations than just the Java SDK) but having those values in enums means every time you want to include a new value, you have to do all new SDK's and specs. That seems like a hassle if your tax team suddenly needs to add GSTONEXPORTS (or some such) and it'd be a breaking change for everyone relying on that enum being exhaustive.

**Comment by SidneyAllen** [comment]
@aschroder - quick clarification... you are using version 2.2.2 of the SDK with the new models for JSON?

I just tested creating a New account and New Bank Account and outputting the response and the individual values for BankAccountType.  No NPE when I test it?

Console output ....

Create a new Account - Name : Bye93590 Description : Foo boo
class Account {
    code: Hello69409
    name: Bye93590
    accountID: be3783fe-a7b4-4081-90d3-92cf8ea97b1a
    type: EXPENSE
    bankAccountNumber: null
    status: ACTIVE
    description: Foo boo
    bankAccountType: null
    currencyCode: null
    taxType: INPUT
    enablePaymentsToAccount: false
    showInExpenseClaims: false
    propertyClass: EXPENSE
    systemAccount: null
    reportingCode: EXP
    reportingCodeName: Expense
    hasAttachments: null
    updatedDateUTC: 2019-01-09T16:55:52.983-08:00
}
Bank Account type: null
class Account {
    code: 1234682
    name: Checking 89062
    accountID: cfa7666b-9b23-4cd5-b79c-efa5759368cf
    type: BANK
    bankAccountNumber: 123486066
    status: ACTIVE
    description: null
    bankAccountType: BANK
    currencyCode: USD
    taxType: NONE
    enablePaymentsToAccount: false
    showInExpenseClaims: false
    propertyClass: ASSET
    systemAccount: null
    reportingCode: ASS
    reportingCodeName: Assets
    hasAttachments: null
    updatedDateUTC: 2019-01-09T16:55:53.310-08:00
}
Bank Account type: BANK

**Comment by SidneyAllen** [comment]
@aschroder - pretty sure you want to update the deprecated XSD Schema files

https://github.com/XeroAPI/XeroAPI-Schemas/blob/master/src/main/resources/XeroSchemas/v2.00/BaseTypes.xsd

Happy to accept a PR to update the old models.

**Comment by aschroder** [comment]
@SidneyAllen I just re-tested with 2.2.2 and it worked! I was on 2.2.1.

Using:
```
 <dependency>
      <groupId>com.github.xeroapi</groupId>
      <artifactId>xero-java</artifactId>
      <version>2.2.2</version>
    </dependency>
```

This SDK is [uses 1.0](https://github.com/XeroAPI/Xero-Java/blob/master/pom.xml#L42) of the API spec, but that doesn't seem to have changed between 2.2.1 and 2.2.2 so I'm not sure why it works now, I'm a bit confused, but happy it is working now regardless - I can dispense with my custom overrides.

Thanks Sid.

**Comment by aschroder** [comment]
@SidneyAllen I'm going to sound like an idiot but this isn't actually fixed from what I can see running 2.2.2 I still hit issues where e.g `SystemAccountEnum` is failing because it has no "None" value for the empty strings in the API response.

**Comment by SidneyAllen** [comment]
@aschroder Turns out you can set an an ENUM in the spec with '' and it will be equal to "EMPTY"

public enum BankAccountTypeEnum {
    BANK("BANK"),
    CREDITCARD("CREDITCARD"),
    PAYPAL("PAYPAL"),
    EMPTY("");
